File: 03_Stack/03_Evaluate_Reverse_Polish_Notation.java
class Solution {
    public int evalRPN(String[] tokens) {
        Map<String, BiFunction<Integer, Integer, Integer>> m = new HashMap<>();

        m.put("+", (a, b) -> a + b);
        m.put("-", (a, b) -> a - b);
        m.put("*", (a, b) -> a * b);
        m.put("/", (a, b) -> a / b);

        Stack<Integer> st = new Stack<>();

        for (String s : tokens) {
            if (!m.containsKey(s))
                st.push(Integer.parseInt(s));
            else {
                int op1 = st.pop();
                int op2 = st.pop();
                st.push(m.get(s).apply(op2, op1));
            }
        }

        return st.pop();
    }
}

File: 03_Stack/03_Evaluate_Reverse_Polish_Notation.cpp
class Solution
{
public:
    int evalRPN(vector<string> &tokens)
    {
        unordered_map<string, function<int(int, int)>> m = {
            {"+", [](int a, int b)
             { return a + b; }},
            {"-", [](int a, int b)
             { return a - b; }},
            {"*", [](int a, int b)
             { return a * b; }},
            {"/", [](int a, int b)
             { return a / b; }}};

        stack<int> st;

        for (string &s : tokens)
        {
            if (!m.count(s))
                st.push(stoi(s));
            else
            {
                int op1 = st.top();
                st.pop();
                int op2 = st.top();
                st.pop();
                st.push(m[s](op2, op1));
            }
        }
        return st.top();
    }
};